### PR TITLE
Avoid exceptions when `count` does not match the number of elements

### PR DESCRIPTION
**Pull request by m7913d**
Fixes issue #735

### CODE REVIEW

**Comment by m7913d** [comment]
@doomlaur Some remarks:
 - Should we add a unit test for this issue?
 - Should we add a CMake options to set the THROW_ON_INVALID_XML define?
 - Should we limit the reserved size to avoid a bad/untrusted (small) file from exhausting our memory?

**Comment by m7913d** [comment]
Note that another alternative is to just remove the throwing code. Tfussel already did this in PR https://github.com/tfussell/xlnt/pull/652.

**Comment by doomlaur** [comment]
@m7913d 
- Good point, adding a unit test is a good idea :+1: I don't have any public (non-confidential) XLSX files containing this issue that I can add as a test to this repository, but I can create a very simple one manually.
- We could add a CMake option for `THROW_ON_INVALID_XML`, or alternatively remove the throwing code. Before fixing the issue and creating the pull request, my first thought was also to simply remove the throwing code. However, this is a correctness issue, so while it's fine for XLNT to be able to open files that are just slightly broken, when writing a serializer (for exporting Excel files), we do absolutely want to ensure absolute correctness (as much as possible, at least). So personally, I see `THROW_ON_INVALID_XML` as a great option for developing XLNT itself, rather than an option that users of the XLNT libraries would use. However, if you have a good use case in mind where the option could also be useful for users of the library and not just for development, please let me know! :smile: For this reason, `THROW_ON_INVALID_XML` is currently an option that can only internally be enabled, although I do agree that it might be possible for other XLNT developers and contributors to oversee it and unwillingly contribute code that could be slightly broken. Another possibility would be to enable `THROW_ON_INVALID_XML` in debug builds but disable it in release builds - however, I'm open for ideas!
- You are right, I unfortunately did not think about the case of completely broken files. This is something that could easily happen in an external software, e.g. use -1 for an invalid `count`, then serialize it as `size_t`, causing an integer underflow and thus causing the `count` to become 18446744073709551615, then causing XLNT to allocate all available memory. Definitely not good, I agree. Maybe we should limit the maximum allocation to something like 100 MB, which is probably way more than anything that could ever be needed.

I can gladly work on these improvements - just let me know if you have any other thoughts on possible improvements, then I can start working :wink:

**Comment by m7913d** [comment]
- Creating a very simple one (that can be opened by Microsoft Excel) is perfect.
 - Ok for me to keep `THROW_ON_INVALID_XML` as an internal option. I wouldn't enable if by default on debug builds, to keep consistency between debug and release build behavior. It may be useful to enable it on some unit tests (in the future). I'm trying to enable the CI tools used by tfussel (AppVeyor and circleci) to run the unit test for the PR, but something is wrong the new account I created on AppVeyor.
 - 100MB seems fine. Maybe we should add it as a constant in e.g. detail/constants.hpp?

No other remarks.

**Comment by doomlaur** [comment]
I agree that a difference between debug and release builds would not be ideal - using unit tests and the CI tools would probably be the best compromise.

Let me know if there's anything I can help with regarding AppVeyor and CircleCI. I don't have experience with either of these tools, but I did use TeamCity in the past.

I'm going to work on the discussed improvements as soon as possible (either today or at least in the next couple of days) :wink:

**Comment by doomlaur** [comment]
Alright, so:

- I limited the maximum number of allocated elements when reading the "count" attribute. Currently, I set the limit to 10000, as I think this is a large enough number to consider more elements as being potentially invalid or maliciously crafted. I did think about limiting it to 100 MB as initially proposed, but wanted to avoid additionally adding complexity by requiring the programmer to always call `sizeof` with the correct size. I think the current solution is good enough :smile: 
- I added a unit test for preventing such issues happen in the future. For that I used a file where I replaced every occurrence of `count` by a broken number.
- By using adding the unit test, I noticed that I forgot one case where an exception could still be thrown (in a `uniqueCount` attribute) and fixed it right away.

Let me know if you have any other comments, and thanks for reviewing! :wink:

File: source/detail/serialization/xlsx_consumer.cpp
Comment (m7913d) [inline source/detail/serialization/xlsx_consumer.cpp:2883]: We can also ifdef this code, as parsing the count without using it is rather a small waste of resources.

File: source/detail/serialization/xlsx_consumer.cpp
Comment (m7913d) [inline source/detail/serialization/xlsx_consumer.cpp:2910]: idem

File: source/detail/serialization/xlsx_consumer.cpp
Comment (m7913d) [inline source/detail/serialization/xlsx_consumer.cpp:2326]: Alternative may be to reserve `xlnt::constants::max_elements_for_reserve()` elements if the requested count is larger than this maximum, in stead of just reserving nothing. Or should we really consider a value larger than max as completely trash?
What do you think?

**Review by m7913d — COMMENTED** [review]
Thanks for the changes!

File: source/detail/serialization/xlsx_consumer.cpp
Comment (doomlaur) [reply source/detail/serialization/xlsx_consumer.cpp:2326]: This is a good question that I also asked myself, but was also not sure whether it necessarily pays off. On the one hand, this would mean that the broken or maliciously crafted files would allocate too much memory unnecessarily. On the other hand, 10000 elements is not an extremely large number, so probably this would never be an issue anyway.

It would be possible to easily achieve what you said by doing something like this:
```cpp
vec.reserve(xlnt::reserve_with_limit(count.get()));
```

Where `xlnt::reserve_with_limit` would do something like this:
```cpp
size_t reserve_with_limit(size_t num_elements)
{
    return std::min(num_elements, xlnt::constants::max_elements_for_reserve());
}
```

This way, we would also make it simple for the caller (would not be much more code).

Then I however asked myself whether this is really necessary or whether I'm overthinking it, so for now I left it as it is. Thinking that most files won't be broken or malicious, it could be a nice performance increase to allocate memory right away in case a very large number of elements is necessary. Due to the fact that the exact way `std::vector::reserve` behaves is not specified in the C++ standard, I now tested the following code:
```cpp
#include <iostream>
#include <vector>

int main()
{
    constexpr size_t max_size = 10'000'000;
    size_t current_capacity = 0;
    size_t capacity_increase_count = 0;
    std::vector<char> vec;
    //vec.reserve(10000);

    for (size_t n = 0; n < max_size; ++n)
    {
        if (vec.capacity() != current_capacity)
        {
            ++capacity_increase_count;
            std::cout << "size: " << vec.size() << " | capacity: " << vec.capacity() << '\n';
            current_capacity = vec.capacity();
        }

        vec.push_back(0);
    }

    std::cout << "\nCapacity increased " << capacity_increase_count << " times.";
}
```

I tested it in the [Compiler Explorer](https://godbolt.org/), then uncommented the line with `reserve`. Results under x86-64:
| Mode | GCC 14.2  | Clang 18.1.0 | MSVC v19.40 VS17.10 |
| ------------- | ------------- | ------------- | ------------- |
| Without | 25 times | 25 times | 41 times |
| 10000 elements | 11 times | 11 times | 19 times |

So I think the wrapper around `max_elements_for_reserve()` I proposed above would make the code simple enough, while also providing a good performance increase even for a larger amount of elements. Please tell me what you think about it, and also feel free to play with the numbers :wink:

Comment (m7913d) [reply source/detail/serialization/xlsx_consumer.cpp:2326]: You may choose, both are fine for me. If we would measure the execution time, the advantage of reserving 10 000 if the needed size is larger is probably less, as the reallocation is more expensive for the larger sizes. If you go for the clipping, I would only prefer a name like `clip_reserve_elements` as the name `reserve_with_limit` gives me the intention that it actually reserves something.

**Comment by m7913d** [comment]
@doomlaur You will have to officially review and merge yourself as I was the creator of this PR.

File: source/detail/serialization/xlsx_consumer.cpp
Comment (doomlaur) [reply source/detail/serialization/xlsx_consumer.cpp:2326]: By the way, due to the fact that the function no longer returns a constant, I thought about creating a new file `utils/limits.hpp` for all kinds of limits (like element reservation limits, or others). Unless you have a better idea, I'll do it like this and merge the PR afterwards :smiley:

Comment (m7913d) [reply source/detail/serialization/xlsx_consumer.cpp:2326]: I don't think this file should be in the include directory (as it is an implementation detail, isn't?). I would rather place it in the same directory of `constants.php` (= detail folder) or create a utils there if desired.

**Review by doomlaur — APPROVED** [review]
It's always awesome to approve my own PRs :sweat_smile:

**Comment by doomlaur** [comment]
@m7913d I still need your approval, as I was the last one that pushed something :wink: Thanks!

File: source/detail/limits.hpp
Comment (m7913d) [inline source/detail/limits.hpp:1]: What should we do with the copyright?
At least updating the year if creating a new file?
Changing Thomas Fussell by XLNT Community?
Or should we keep Thomas Fussell and just add an extra copyright line? Or should we only keep his name in existing files?

Also AUTHORS may need some update in the future, not sure if we need to add all are names or just XLNT Community and refer to the git committers. The list in authors are currently really small compared to the numbers of people that actually contributed at least a single commit.

**Review by m7913d — COMMENTED** [review]
Look goods, just some considerations about the copyright notice.

**Comment by m7913d** [comment]
@doomlaur Github doesn't allow me to approve, as the PR creator can never approve. LOL

We need more reviewers for this pull request.

File: source/detail/limits.hpp
Comment (doomlaur) [reply source/detail/limits.hpp:1]: I think the standard procedure is to keep the original author and add an extra copyright line, yes. But this is without having researched the actual recommendations, but just based on what I saw in other projects. The reason why I left it like this - for now - is because [my own XLNT repository](https://github.com/doomlaur/xlnt) was forked from the original one, and [Pull Request #736](https://github.com/tfussell/xlnt/pull/736) should still be merged - if that will ever happen - to the original repository. In other words, in my (slightly unresearched) opinion, it is correct for my PR to still contain Thomas Fussell's copyright. Of course, as I said, I'm not an expert on this and we should probably research it better before deciding on how to go forward.

As for new Pull Requests, I agree, we should modify the copyright, in my opinion by adding an additional line. TODO: research this topic.

**Comment by doomlaur** [comment]
@m7913d Haha LOL :laughing: For me, GitHub says:
```quote
New changes require approval from someone other than doomlaur because they were the last pusher.
```

I like the rule that a second maintainer has to approve, but can we somehow override this rule just for this PR, as it currently doesn't allow us to merge it? :sweat_smile:  Otherwise the only other option I see is that we would have to close this PR and I would need to open a new one myself, then I would need your approval.

File: source/detail/limits.hpp
Comment (m7913d) [reply source/detail/limits.hpp:1]: Ok, I agree that on existing files keeping Thomas Fussell is a good idea.

Let's merge this PR.

**Comment by m7913d** [comment]
@doomlaur I temporarily disabled the requirements that someone needed to review your latest commit, performed the merge and reenabled it.

In the future, it is a good idea to not create a PR for code written by the other one.

**Comment by m7913d** [comment]
All PRs and Issues closed. Good job!

**Pull request merged**
